solidity-bindgen-macros/src/abi_gen.rs

```diff
@@ -105,51 +106,62 @@ pub fn fn_from_abi(abi: &Abi, span: Span) -> TokenStream {
                     #name: #t
                 }
             });
 
-            let body = {
-                let params = function.inputs.iter().enumerate().map(|(i, param)| {
-                    let name = Ident::new(&to_rust_name("input", &param.name, i), span);
-                    quote! { #name }
-                });
-                if function.constant.unwrap_or_default() {
-                    quote! {
-                        self.contract.query(#eth_name, (#(#params),*)).await
-                    }
-                } else {
-                    // Non-pure functions need to use call_with_verifications instead of query,
-                    // and payable functions may yet need something else
-                    quote! {
-                        self.contract.non_pure_todo(#eth_name, (#(#params),*)).await
-                    }
-                }
+            let params = function.inputs.iter().enumerate().map(|(i, param)| {
+                let name = Ident::new(&to_rust_name("input", &param.name, i), span);
+                quote! { #name }
+            });
+            let params = if function.inputs.len() == 1 {
+                quote! { #(#params)* }
+            } else {
+                quote! { (#(#params),*) }
             };
 
-            let ok = match function.outputs.len() {
-                0 => quote! { ::solidity_bindgen::internal::Empty },
-                1 => {
-                    let t = param_type(&function.outputs[0].r#type);
-                    quote! { #t }
-                }
-                _ => {
-                    let types = function.outputs.iter().map(|o| {
-                        let t = param_type(&o.r#type);
+            let transaction = matches!(
+                function.state_mutability,
+                StateMutability::Nonpayable | StateMutability::Payable
+            );
+            let method = if transaction { "send" } else { "call" };
+            let method = Ident::new(method, Span::call_site());
+
+            let ok = if transaction {
+                // Despite information in the ABIs to the contrary, there aren't
+                // really outputs for web3 send fns. The outputs that are
+                // available aren't returned by these APIs, but are only made
+                // available to contracts calling each other. 🤷
+                //
+                // All you can get is a receipt. So, the way to get something
+                // like a return value would be to check for events emitted or
+                // to make further queries for data.
+                quote! { ::web3::types::TransactionReceipt }
+            } else {
+                match function.outputs.len() {
+                    0 => quote! { ::solidity_bindgen::internal::Empty },
+                    1 => {
+                        let t = param_type(&function.outputs[0].r#type);
                         quote! { #t }
-                    });
+                    }
+                    _ => {
```

Comment:
Do we need the separate match arms for `1` and `_`?

Reply:
> Do we need the separate match arms for `1` and `_`?

These differ in whether we output a tuple. No need for a 1 item tuple or an unnecessary parens warning.

Reply:
👍